### PR TITLE
DEV-1750: lift time_shift × cross-model guard + shifted-CTE fragment join discovery

### PR DESCRIPTION
**Pull request by ZmeiGorynych**
Closes DEV-1750. Spun out of DEV-1745 (PR 2 of 6), deliberately **outside** the ratified DEV-1742 6-PR chain — it does not become "PR 7" and must not block the umbrella merging forward.

## What

The Mode-A custom-aggregation template mechanism discovers the join paths its `{value}`/`{param}` fragments cross by scanning them. That scan lived on the host and (since DEV-1745) the `_cm_` CTE paths, but never the shifted (`time_shift`) CTE — and the whole `time_shift`/`consecutive_periods` × cross-model combination was rejected by a blanket `stage 7b.15e` `NotImplementedError`.

**Part 1 — shifted-CTE join discovery through the one door.** `_emit_time_shift_ctes_for_planned` now registers the inner aggregate's source (widened to path-bearing), positional column args (skipping the DEV-1526 residual), column kwargs, and — the missing piece — the template fragments via `_register_fragment_kwarg_joins`, the same helper the host base and `_cm_` CTE use. So `time_shift(amount:wscaled_sum, -1)` with a default `w='customers__regions.weight'` pulls `LEFT JOIN customers` / `LEFT JOIN regions` into the shifted CTE's FROM instead of emitting `SUM(orders.amount * customers__regions.weight)` with no join (SQL no database binds).

**Part 2 — narrow the guard, add the arms.** The cross-model transform chain gained the same window / `time_shift` / `consecutive_periods` Kahn arms the local chain runs, dispatching temporal ops to the shared per-op emitters (so join discovery can't drift between the two chains — full loop unification is [DEV-1799](https://linear.app/motley-ai/issue/DEV-1799)). The blanket guard is replaced by `_guard_target_grain_time_shift`, which fires **only** for a `time_shift` whose inner aggregate is a **target-grain** cross-model aggregate — detected by **plan ownership** (`CrossModelAggregatePlan.cte_root_model is None`), not formula text — because host-rooted re-aggregation there would multiply target rows through the 1:N join.

## Scope

- **Renders:** `time_shift`/`consecutive_periods` over a local or host-rooted inner beside a cross-model aggregate; `change`/`change_pct` over a local inner. `consecutive_periods` is lifted entirely (it reads a materialised alias, never re-aggregates).
- **Stays a loud, specific error:** `time_shift` over a target-grain cross-model inner (narrowed 7b.15e); `time_shift` over a ranked `first`/`last` inner (the shifted CTE can't reproduce the ROW_NUMBER ranking).
- **Out of scope → [DEV-1800](https://linear.app/motley-ai/issue/DEV-1800):** `change`/`change_pct` over a *cross-model* inner hits a separate pre-existing `RenderContextMissingFacilityError` (combined-arithmetic-over-transform), pinned here as a loud error.

## Tests

- `test_dev1750_execution.py` — SQLite **and** DuckDB ground truth: shapes a/b, cp, change/change_pct, NULL-dim null-safe survival, `date_range` frame-bound omission, and sibling-protection under a genuine 1:N (`line_items`) fan-out.
- `test_dev1750_guard_ownership.py` — planner-level pin that the guard reads `cte_root_model`, not formula text.
- `test_dev1750_shifted_fragment_joins.py` — the shifted CTE joins the fragment hops; base doesn't.
- `test_dev1750_guard_lift.py` — routing + narrowed guard fires before the emitter.
- `test_dev1750_golden_sql.py` — per-dialect emission (incl. Postgres/T-SQL/BigQuery).

Full non-integration suite green (11898 passed); `ruff check slayer/ tests/` clean; no emitted SQL changed for any pre-existing shape.

## Summary by CodeRabbit

* **New Features**
  * Added support for `time_shift` and `consecutive_periods` across local and cross-model aggregates.
  * Improved temporal calculations across aggregate fragments and joined data.
* **Bug Fixes**
  * Improved shifted-query alignment, join handling, alias safety, and protection against join fan-out.
  * Added clearer handling for unsupported aggregate configurations.
* **Documentation**
  * Updated architecture and CLI documentation to clarify supported transformations, limitations, and options.
* **Tests**
  * Expanded SQL generation and execution coverage across supported database dialects.

### CODE REVIEW

File: slayer/core/models.py [hunk 1]
```diff
@@ -741,6 +764,12 @@ class DatasourceConfig(BaseModel):
     # When unset, BigQuery falls back to Application Default Credentials
     # (``GOOGLE_APPLICATION_CREDENTIALS`` env var or attached compute identity).
     credentials_json: str | None = Field(default=None, repr=False)
+    # BigQuery-specific. A Google OAuth authorized-user grant as JSON, in the
+    # shape ``Credentials.from_authorized_user_info`` consumes — the per-end-user
+    # auth path, so queries run with that user's permissions. Mutually exclusive
+    # with ``credentials_json``, which cannot carry a grant: the driver feeds it
+    # to ``from_service_account_info``.
+    oauth_credentials_json: str | None = Field(default=None, repr=False)
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/core/models.py:772]: @coderabbitai Already enforced at engine-build time: `BigqueryDialect.build_engine` raises a "mutually exclusive" `ValueError` when both `credentials_json` and `oauth_credentials_json` are set (`slayer/sql/dialects/bigquery.py:203-209`), and `tests/dialects/test_bigquery.py::test_build_engine_rejects_both_credential_kinds` constructs a `DatasourceConfig` with both set and asserts exactly that. `DatasourceConfig` deliberately permits the intermediate state, so a model-level validator would break that established test boundary. Independently confirmed.

File: tests/dialects/test_bigquery.py [hunk 2]
```diff
@@ -821,3 +823,295 @@ async def test_bigquery_dry_run_does_not_decode_data_rows() -> None:
...
+    info = {  # NOSONAR(S2068) — test fixture; placeholder grant, not real credentials
+        "type": "authorized_user",
+        "client_id": "cid.apps.googleusercontent.com",
+        "client_secret": "csecret",
+        "refresh_token": "rtok-alice",
+        "token": "access-token-1",
+        "token_uri": "https://oauth2.googleapis.com/token",
+    }
+    info.update(overrides)
+    return info
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply tests/dialects/test_bigquery.py:842]: @coderabbitai No secret scanner (gitleaks/Betterleaks) is configured in this repo's CI (nothing under `.github/`), so there is no gate to allowlist against, and the fixture already carries `# NOSONAR(S2068)`. If a secret-scanning gate is added later, the inline allow marker can be added then.

File: tests/dialects/test_identifier_fit.py [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+``substitute_quoted`` is the two-phase write side. Emission: ``test_dev1756_identifier_length.py``.
+"""
+
+from __future__ import annotations
+
+import hashlib
+import os
+import re
+import subprocess
+import sys
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply tests/dialects/test_identifier_fit.py:14]: @coderabbitai `sys.executable` is intentional here: it launches the SAME interpreter already running pytest (under Poetry) while varying only `PYTHONHASHSEED` — exactly the isolation this digest-determinism check needs. `poetry run python` would re-resolve the environment and could select a different interpreter. The "use `poetry run`" rule targets developer-invoked shell commands, not an in-test subprocess of the current interpreter. Independently confirmed.

File: slayer/engine/ingestion.py [hunk 4]
```diff
@@ -748,123 +1054,440 @@ def introspect_table_to_model(
...
+            "view",
+        )
+        _add(
+            _safe_object_names(
+                accessor_name="get_materialized_view_names",
+                inspector=inspector,
+                schema=schema,
+            ),
+            "materialized_view",
+        )
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/ingestion.py]: @coderabbitai Real edge case, but the current "first-seen (table→view→matview order) wins the kind" behavior is asserted intentional by `tests/test_ingestion_views.py::TestDialectTolerance::test_name_returned_as_both_table_and_view_appears_once` ("first-seen wins as a table") and `test_matview_also_listed_as_view_appears_once`. Making the view/matview listing authoritative is a deliberate behavior change that would rewrite those tests, so it's being handled as a separate design decision rather than folded into this (DEV-1750) PR, which is scoped to the transform-chain work.

File: slayer/engine/ingestion.py [hunk 5]
```diff
@@ -1254,18 +2051,57 @@ def _friendly_db_error(exc: Exception) -> str:
...
 
+def _get_schemas(ds: DatasourceConfig) -> list[str]:
+    """List a datasource's schemas. Best-effort; empty means no hint."""
+    try:
+        from slayer.sql import engine_factory
+        engine = engine_factory.get_engine(ds.resolve_env_vars())
+        inspector = sa.inspect(engine)
+        return inspector.get_schema_names()
+    except Exception:  # noqa: BLE001 — hint-only and never fatal
+        return []
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/ingestion.py:2250]: @coderabbitai `engine_factory.get_engine` does return a cached, shared engine, so the observation is valid. But `ingest_datasource_report` disposing it is asserted intentional by `tests/test_ingestion_name_sanitize.py::TestEngineDisposal`, which pins the dispose call and its error handling. Removing the dispose breaks those tests, so changing this engine lifecycle is a deliberate design decision tracked separately, not folded into this PR.

File: slayer/sql/generator.py [hunk 6]
```diff
@@ -8739,6 +8895,7 @@ def generate_planned_stages(
     *,
     bundle,
     dialect: str = "postgres",
+    projection_aliases: "Sequence[str]" = (),
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/sql/generator.py:8855]: @coderabbitai Confirmed real — the query-backed backing SQL does emit over-limit projection aliases on bounded dialects. But the one-line fix (passing `projection_aliases=projection_result_keys(...)`) is insufficient: length-fitting the backing SQL then makes `build_flat_rename_wrapper` raise `ValueError: rendered output columns [...fitted...] do not match the expected schema [...canonical...]`, because the wrapper matches the backing SQL's output columns against the canonical (unfitted) stage schema. A correct fix must also teach the wrapper the fitted↔canonical mapping — a larger change tracked separately rather than a one-liner here.

File: DECISIONS.md [hunk 7]
```diff
@@ -117,3 +117,4 @@ implementation detail. Include issue refs when known.
...
...and never re-aggregates, so it has no target-grain failure mode (executed for the target-grain-inner shape). A `time_shift` over a ranked `first`/`last` aggregate raises a distinct loud 7b.15e (the shifted CTE re-aggregates flatly and cannot reproduce the ROW_NUMBER ranking; the crossing time arg is still registered for Law-1 totality). **Scope boundary (author-ratified):** `change`/`change_pct` render over a **local/host-rooted** inner (the chain's existing outer arithmetic-materialization step handles the subtraction) but over a **cross-model** inner they hit a separate pre-existing `RenderContextMissingFacilityError` in the combined SELECT (arithmetic-over-transform rendered before the transform materialises) — out of scope, tracked as **DEV-1800** and pinned as a loud, specific error. Execution ground truth (SQLite + DuckDB) for shapes a/b, cp, change/change_pct, NULL-dim null-safe survival, `date_range` frame-bound omission, and sibling-protection under a genuine 1:N (`line_items`) fan-out; per-dialect emission (incl. Postgres/T-SQL/BigQuery) pinned in `tests/golden/dev1750_sql_baseline.json`. Full non-integration suite green; no emitted SQL changed for any pre-existing shape.
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply DECISIONS.md]: @coderabbitai Condensed to a 3-sentence decision + rationale in commit 26555697.

File: docs/concepts/references.md [hunk 8]
```diff
@@ -135,6 +136,16 @@ Four consequences worth knowing:
     {"filters": ["concat(ifnull(first_name, ''), ifnull(last_name, '')) = 'AdaLovelace'"]}
     ```
 
+* **`greatest` / `least` NULL handling is backend-specific.** They pass through
+  to each backend's native form, and those forms disagree on `NULL`: Postgres,
+  DuckDB, SQL Server and Snowflake *ignore* `NULL` arguments (returning `NULL`
+  only when every argument is `NULL`), while SQLite (scalar `MAX`/`MIN`), MySQL,
+  ClickHouse and BigQuery *propagate* `NULL` (any `NULL` argument makes the
+  result `NULL`). SLayer does not normalise this — wrap arguments in
+  `ifnull(...)` / `coalesce(...)` if you need one behaviour on every backend.
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply docs/concepts/references.md:146]: @coderabbitai Fixed in b00870e2 — Snowflake moved to the NULL-*propagating* group (its GREATEST/LEAST return NULL on any NULL arg; GREATEST_IGNORE_NULLS/LEAST_IGNORE_NULLS skip nulls).

File: docs/concepts/schema-drift.md [hunk 9]
```diff
@@ -103,11 +103,17 @@ be no-ops once the model is gone).
...
+* **CLI.** `slayer validate-models [--datasource X] [--model M]
+  [--format text|json] [--force-clean] [--yes]`. Without `--force-clean`,
+  prints the diff and exits 0. `--force-clean` prompts (or skips with
+  `--yes`), applies via `apply_drift_deletes`, and exits non-zero on
+  per-entry errors or non-empty residual drift. `--model` scopes both the
+  report and the applied deletes. Unlike the engine method, an unscoped CLI
+  run validates each datasource explicitly and exits 1 if any of them fails,
+  rather than dropping it from the result. `--cardinality` adds an opt-in
+  [join-arity profiling](models.md#join-cardinality) pass, which runs after
+  any `--force-clean` apply. See the
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply docs/concepts/schema-drift.md:115]: @coderabbitai Fixed in b00870e2 — added `[--cardinality]` to the validate-models synopsis.

File: docs/reference/cli.md [hunk 10]
```diff
@@ -93,8 +95,157 @@ slayer ingest --datasource my_postgres --exclude migrations,django_session
...
+
+ELT and migration tooling writes bookkeeping tables into the same schema as
+your data. These are modelled with `hidden: true` and listed under `Hidden`:
+
+```text
+Hidden (3) — recognised ELT/migration internals (excluded from models_summary; still queryable by name):
+  - _dlt_loads: dlt
+  - _dlt_pipeline_state: dlt
+  - _dlt_version: dlt
+  --surface-internals ingests NEW internals visible; use edit_model("<model>", data_source="my_postgres", hidden=false) to unhide an existing one.
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply docs/reference/cli.md]: @coderabbitai Fixed in b00870e2 — `hidden=False` (Python literal) in both the cli.md and mcp.md `edit_model` examples.

File: slayer/engine/ingestion.py [hunk 11]
```diff
@@ -387,39 +436,256 @@ def _compute_transitive_closure(graph: dict[str, set[str]], source: str) -> set[
...
+        else:
+            # Only the bare-inspector path needs normalizing;
+            # _safe_get_pk_constraint already guarantees a mapping.
+            pk = inspector.get_pk_constraint(table_name=table_name, schema=schema)
+            if not isinstance(pk, dict):
+                return []
+    except Exception:
+        return []
+    cols = pk.get("constrained_columns")
+    return [list(cols)] if cols else []
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/ingestion.py:525]: @coderabbitai Fixed in 4c3f506f — the bare-inspector path now guards `constrained_columns` with `isinstance(cols, list)` so a bare string can't split into a per-character key-set; regression test added (Codex-reviewed).

File: slayer/engine/query_engine.py [hunk 12]
```diff
@@ -2425,6 +2455,254 @@ async def apply_drift_deletes(
...
+            exp.select(*[c.copy() for c in cols])
+            .from_(tbl.copy())
+            .where(predicate.copy())
+            .distinct()
+        )
+        dist_q = exp.select(count_star.copy()).from_(inner.subquery("d"))
+        return (
+            rows_q.sql(dialect=sqlglot_name, identify=True),
+            dist_q.sql(dialect=sqlglot_name, identify=True),
+        )
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/query_engine.py:2656]: @coderabbitai Fixed in b00870e2 — `exp.to_table(table, dialect=sqlglot_name)`, matching the cache.py pattern.

File: slayer/mcp/server.py [hunk 13]
```diff
@@ -82,48 +158,40 @@ def _test_connection(ds: DatasourceConfig) -> tuple[bool, str]:
...
     try:
         from slayer.sql import engine_factory
         sa_engine = engine_factory.get_engine(ds.resolve_env_vars())
         inspector = sa.inspect(sa_engine)
-        tables = inspector.get_table_names(schema=schema_name)
-        return sorted(tables), None
+        objects = list_ingestable_objects(
+            inspector=inspector, schema=schema_name, include_views=True
+        )
+        return sorted(o.name for o in objects), None
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/mcp/server.py]: @coderabbitai Fixed in 4c3f506f — `_fetch_tables` now returns `IngestableObject` (name + kind) and `describe_datasource` labels non-table objects `(view)` / `(materialized_view)`. (The update-path `widened_columns` / `kind_change` were already rendered.) Test updated to assert kind is preserved.

File: DECISIONS.md [hunk 14]
```diff
@@ -140,3 +140,4 @@ implementation detail. Include issue refs when known.
...
+- 2026-08-18 — `time_shift` / `consecutive_periods` over a cross-model aggregate now render (DEV-1750, spun out of DEV-1745, outside the DEV-1742 6-PR chain). The blanket 7b.15e guard is narrowed: the shifted CTE discovers its inner aggregate's template-fragment and positional-arg joins through the same scope door the host / `_cm_` paths use, and the host and cross-model transform chains share one set of per-op step emitters. Only a `time_shift` whose inner is a TARGET-GRAIN cross-model aggregate stays guarded — re-aggregating it host-rooted would multiply target rows through the 1:N join; `consecutive_periods` reads a materialised alias and never re-aggregates, so it is exempt.
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply DECISIONS.md]: @coderabbitai Fixed in b00870e2 — re-dated the DEV-1750 entry to its actual append date (2026-08-22) so the flat-chronological "newest at bottom" convention holds; existing entries were not reordered.

File: slayer/engine/ingestion.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/ingestion.py]: Fixed in 371635e6. `list_ingestable_objects` now resolves the kind most-specific-wins (matview > view > table) from the prefetched view/matview listings, so a view returned by `get_table_names()` is stamped `view` and `source_kind` matches its documented meaning. Emission order and first-seen dedup are unchanged (name-collision reservation still needs the stable tables→views→matviews scan). The two tests that pinned the old behavior were updated, and regression tests were added for the table+view dup, the view+matview dup, and a name in all three listings. Rationale recorded in DECISIONS.md.

File: slayer/engine/ingestion.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/ingestion.py:2250]: The observation is correct — `get_engine` returns a cached, shared engine — but the dispose turns out to be load-bearing, so we're keeping it deliberately (371635e6 rewrites the comment to say so; full rationale in DECISIONS.md).

Empirical check: without the dispose, a same-process `duckdb.connect(file)` after a scan fails with "Can't open a connection to same database file with a different configuration than existing connections", which breaks the ingest-then-external-write DuckDB flow (pinned by `tests/test_ingestion_comments.py::test_new_commented_column_arrives_with_description`). `engine.dispose()` only closes pooled connections — the shared Engine object stays valid and reconnects lazily — so the cost is pool churn on a rare one-shot admin path, which we accept. Alternatives (a private uncached engine per scan, factory-side `invalidate_engine`) were considered and rejected.

File: slayer/sql/generator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/sql/generator.py:8855]: Fixed in 371635e6 — confirmed real, though the suggested one-liner alone breaks `build_flat_rename_wrapper` (it matches rendered output columns against the canonical stage schema, so fitting the render made the match raise). The full fix: the render gets `projection_aliases`, and the wrapper accepts the same list — decoding fitted/mangled rendered names back to canonical for the schema match and length-fitting its own emitted flat aliases via `alias_rewrite_map` (which also guards collisions). The virtual model bridges with `Column.name` canonical / `Column.sql` fitted, so outer queries treat it like a regular model with a long column name and the existing DEV-1756 write/read machinery applies downstream. Regression tests cover postgres (63B), T-SQL and BigQuery (fit+mangle), an under-limit byte-identical golden, the wrapper-level decode/fit/collision paths, and an outer query through the expanded model. Design note in DECISIONS.md.
